### PR TITLE
Votes cancellation & Staking V2

### PR DESCRIPTION
**Pull request by dovgopoly**

### CODE REVIEW

File: contracts/libs/gov/gov-pool/GovPoolStaking.sol [hunk 1]
```diff
@@ -14,183 +13,197 @@ library GovPoolStaking {
...
+
+        doBeforeRestake(micropoolPair, proposalIds, delegatee);
+
+        _claim(micropoolPair[true], proposals, proposalIds);
+        _claim(micropoolPair[false], proposals, proposalIds);
     }
 
-    function unstake(IGovPool.MicropoolInfo storage micropool, address delegatee) external {
-        _recalculateStakingState(micropool, delegatee, true);
+    function doBeforeRestake(
```
Comment (Arvolear) [inline contracts/libs/gov/gov-pool/GovPoolStaking.sol:64]:
```suggestion
    function beforeRestake(
```

File: contracts/libs/gov/gov-pool/GovPoolStaking.sol [hunk 2]
```diff
@@ -14,183 +13,197 @@ library GovPoolStaking {
...
+        uint256[] calldata proposalIds,
+        address delegatee
+    ) public {
+        _doBeforeRestake(micropoolPair[true], proposalIds, delegatee);
+        _doBeforeRestake(micropoolPair[false], proposalIds, delegatee);
     }
 
-    function updateStakingCache(
-        IGovPool.MicropoolInfo storage micropool,
+    function doAfterRestake(
```
Comment (Arvolear) [inline contracts/libs/gov/gov-pool/GovPoolStaking.sol:73]:
```suggestion
    function afterRestake(
```

Comment (dovgopoly) [reply contracts/libs/gov/gov-pool/GovPoolStaking.sol]: done

File: contracts/libs/gov/gov-pool/GovPoolStaking.sol (hunk 1 above)
Comment (dovgopoly) [reply contracts/libs/gov/gov-pool/GovPoolStaking.sol]: done

File: contracts/libs/gov/gov-pool/GovPoolStaking.sol [hunk 3]
```diff
@@ -15,6 +15,7 @@ library GovPoolStaking {
     using Math for uint256;
 
     event StakingRewardClaimed(address user, address token, uint256 amount);
+    event Joined(address user, uint256 proposalId);
```
Comment (Arvolear) [inline contracts/libs/gov/gov-pool/GovPoolStaking.sol:18]:
```suggestion
    event Joined(address user, uint256 proposalId, address delegatee);
```

File: contracts/gov/validators/GovValidators.sol [hunk 4]
```diff
@@ -147,6 +147,7 @@ contract GovValidators is IGovValidators, OwnableUpgradeable {
 
         addressVoted[proposalId][isInternal][msg.sender] += amount;
 
+        // TODO: use library for this?
```
Comment (Arvolear) [inline contracts/gov/validators/GovValidators.sol:150]: Please remove the comment

File: contracts/libs/gov/gov-pool/GovPoolVote.sol [hunk 5]
```diff
@@ -22,6 +22,14 @@ library GovPoolVote {
         bool isVoteFor
     );
 
+    event VotesCancelled(
```
Comment (Arvolear) [inline contracts/libs/gov/gov-pool/GovPoolVote.sol:25]:
```suggestion
    event VotesCanceled(
```

File: contracts/libs/gov/gov-pool/GovPoolVote.sol [hunk 6]
```diff
@@ -22,6 +22,14 @@ library GovPoolVote {
...
+        uint256 proposalId,
+        address sender,
+        uint256 personalCanceledReward,
+        uint256 delegatedCanceledReward,
```
Comment (Arvolear) [inline contracts/libs/gov/gov-pool/GovPoolVote.sol:29]:
```suggestion
        uint256 personalVote,
        uint256 delegatedVote,
```

File: contracts/libs/gov/gov-pool/GovPoolVote.sol [hunk 7]
```diff
@@ -87,6 +95,56 @@ library GovPoolVote {
...
+        IGovPool.VoteInfo storage voteInfo = voteInfos[proposalId][msg.sender][isMicropool];
+
+        (IGovPool.VoteOption storage voteOption, ) = _voteOptions(voteInfo, isVoteFor);
+
+        require(voteOption.tokensVoted >= voteAmount, "Gov: not enough tokens");
+
+        _saveTokenResult(core, voteInfo, voteAmount, isVoteFor, false);
+
+        for (uint256 i; i < nftIds.length; i++) {
+            require(voteOption.nftsVoted.remove(nftIds[i]), "Gov: NFT was not voted");
```
Comment (Arvolear) [inline contracts/libs/gov/gov-pool/GovPoolVote.sol:125]:
```suggestion
            require(voteOption.nftsVoted.remove(nftIds[i]), "Gov: NFT didn't vote");
```

File: contracts/libs/gov/gov-pool/GovPoolVote.sol [hunk 8]
```diff
@@ -239,10 +289,17 @@ library GovPoolVote {
         bool useDelegated,
         bool isVoteFor
     ) internal returns (uint256 voteAmount) {
-        EnumerableSet.UintSet storage votedNfts = _votedNfts(voteInfo, isVoteFor);
+        (
+            IGovPool.VoteOption storage voteOption,
+            IGovPool.VoteOption storage voteOptionAggainst
```
Comment (Arvolear) [inline contracts/libs/gov/gov-pool/GovPoolVote.sol:294]:
```suggestion
            IGovPool.VoteOption storage voteOptionAgainst
```

File: contracts/libs/gov/gov-pool/GovPoolVote.sol [hunk 9]
```diff
@@ -87,6 +95,56 @@ library GovPoolVote {
...
+    function cancelVotes(
+        mapping(uint256 => IGovPool.Proposal) storage proposals,
+        mapping(uint256 => mapping(address => mapping(bool => IGovPool.VoteInfo)))
+            storage voteInfos,
+        uint256 proposalId,
+        bool isMicropool,
+        uint256 voteAmount,
+        uint256[] calldata nftIds,
+        bool isVoteFor
+    ) external returns (uint256 reallocatedReward) {
```
Comment (Arvolear) [inline contracts/libs/gov/gov-pool/GovPoolVote.sol:107]:
```suggestion
    ) external returns (uint256 canceledReward) {
```

File: contracts/gov/GovPool.sol [hunk 10]
```diff
@@ -248,14 +248,71 @@ contract GovPool is
...
+        uint256 reward = _proposals.cancelVotes(
+            _voteInfos,
+            proposalId,
+            false,
+            voteAmount,
+            voteNftIds,
+            isVoteFor
+        );
+
+        if (proposalId != 0) {
```
Comment (Arvolear) [inline contracts/gov/GovPool.sol:274]: Add requires?

File: contracts/gov/validators/GovValidators.sol (hunk 4 above)
Comment (dovgopoly) [reply contracts/gov/validators/GovValidators.sol:150]: done

File: contracts/libs/gov/gov-pool/GovPoolVote.sol (hunk 5 above)
Comment (dovgopoly) [reply contracts/libs/gov/gov-pool/GovPoolVote.sol:25]: done

File: contracts/libs/gov/gov-pool/GovPoolVote.sol (hunk 6 above)
Comment (dovgopoly) [reply contracts/libs/gov/gov-pool/GovPoolVote.sol:29]: done

File: contracts/libs/gov/gov-pool/GovPoolVote.sol (hunk 7 above)
Comment (dovgopoly) [reply contracts/libs/gov/gov-pool/GovPoolVote.sol:125]: done

File: contracts/libs/gov/gov-pool/GovPoolVote.sol (hunk 8 above)
Comment (dovgopoly) [reply contracts/libs/gov/gov-pool/GovPoolVote.sol:294]: done

File: contracts/libs/gov/gov-pool/GovPoolVote.sol (hunk 9 above)
Comment (dovgopoly) [reply contracts/libs/gov/gov-pool/GovPoolVote.sol:107]: done

**Pull request closed without merging**